Support new FNL/FNA function format

C2thehris opened the pull request:
Per the [LCov man pages](https://github.com/linux-test-project/lcov/blob/5186f7394417292537d17b9a9a30f8c06bedc16c/man/geninfo.1#L1430), there are two fields which replace the obsolete "FN" field:

FNL:\<index\>,\<line number of function start\>\[,\<line number of function end\>\]
FNA:\<index\>,\<execution count\>,\<name\>

This PR adds support for these fields.

friedemannsommer commented on src/constants.ts:
```diff
@@ -32,42 +32,50 @@ export enum Variant {
      */
     FilePath = 6,
+    /**
+     * Call count for a given function alias.
```
```suggestion
     * Call count for a given function alias, identified by index.
```

friedemannsommer commented on src/lib/handle-result.ts:
```diff
@@ -92,11 +106,30 @@ export function createSection(entry?: TestNameEntry): SectionSummary {
 ): void {
     const functionSummary = functionMap.get(entry.name)
+    if (entry.variant === Variant.FunctionAlias) {
+        // biome-ignore lint/style/noNonNullAssertion: Leader always comes before aliases of a function
+        const functionLeader = functionLeaders.get(entry.index)!
```
This may be technically correct, but it would be better to make sure it doesn't crash just because some generator didn't use the expected order.

It would be great if the new implementation would fallback gracefully like the current one does for `FN` fields.

friedemannsommer commented on src/lib/handle-result.ts:
```diff
@@ -92,11 +106,30 @@ export function createSection(entry?: TestNameEntry): SectionSummary {
+        if (functionSummary === undefined) {
+            const summary = {
+                hit: entry.hit,
+                line: functionLeaders.get(entry.index)?.lineStart || 0,
```
Since `0 || 0` will use the _else_ branch even though there is an actual value, consider using the [`??`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) operator to use `0` only when the value is absent (i.e.: undefined).

```suggestion
                line: functionLeaders.get(entry.index)?.lineStart ?? 0,
```

friedemannsommer commented on src/lib/handle-result.ts:
```diff
@@ -26,11 +35,15 @@ export function handleResult(entry: EntryVariants, functionMap: FunctionMap, sec
+            functionLeaders.clear()
             return true
+        case Variant.FunctionLeader:
+            functionLeaders.set(entry.index, entry)
```
Please move this insert operation to the `createUpdateFunctionSummary` function so that it can either insert or update the `Map` value if the expected order doesn't match the given one.

friedemannsommer commented on src/lib/handle-result.ts:
```diff
@@ -12,12 +14,19 @@ import type {
+): boolean {
     switch (entry.variant) {
         case Variant.TestName:
             // creating a new test module, to prevent name conflicts simply clear the current FunctionMap
```
It would be good to change this comment as well, either to mention the new `Map` as well, or (preferably) to remove the explicit mention and just describe why these maps need to be cleared.

friedemannsommer commented on src/lib/handle-result.ts:
```diff
@@ -26,11 +35,15 @@ export function handleResult(entry: EntryVariants, functionMap: FunctionMap, sec
         case Variant.EndOfRecord:
             // we've left the test module, to prevent name conflicts clear the current FunctionMap
```
It would be good to change this comment as well, either to mention the new `Map` as well, or (preferably) to remove the explicit mention and just describe why these maps need to be cleared.

friedemannsommer commented on src/tests/transform-result.spec.ts:
```diff
@@ -1134,3 +1168,197 @@ describe('transformComment', (): void => {
+
+describe('transformFunctionLeader', (): void => {
+    const testData: TestData<Variant.FunctionLeader, FunctionLeaderEntry> = [
+        [
+            {
+                done: false,
+                incomplete: false,
+                value: null,
+                variant: Variant.FunctionLeader
+            },
+            {
+                done: false,
+                index: 0,
+                lineEnd: 0,
+                lineStart: 0,
+                variant: Variant.FunctionLeader
+            }
+        ],
```
`value` can (or rather _should_) never be `null`, so having a test for it seems unnecessary.

friedemannsommer commented on src/tests/transform-result.spec.ts:
```diff
@@ -1134,3 +1168,197 @@ describe('transformComment', (): void => {
+
+describe('transformFunctionAlias', (): void => {
+    const testData: TestData<Variant.FunctionAlias, FunctionAliasEntry> = [
+        [
+            {
+                done: false,
+                incomplete: false,
+                value: null,
+                variant: Variant.FunctionAlias
+            },
+            {
+                hit: 0,
+                done: false,
+                index: 0,
+                name: '',
+                variant: Variant.FunctionAlias
+            }
+        ],
```
`value` can (or rather _should_) never be `null`, so having a test for it seems unnecessary.

friedemannsommer reviewed (changes requested):
Hi @C2thehris, thanks for contributing this implementation.
Please read the comments below.
If you have any questions, please let me know.

friedemannsommer commented:
>     1. Adding support for this adds some additional complexity in createUpdateFunctionSummary. Maybe I should break this into a separate function?
> 
>     2. My implementation still uses FunctionLeaderEntry, but I have added an "aliases" property to it. Should I define a new interface for this purpose instead? FunctionLeaderEntry has some properties which don't really make sense here (like "variant" and "done").

1. Sounds like a good idea to separate these two.
2. I agree, it would be better to define an internal `interface` for this purpose.

Thank you for implementing the changes so quickly.

C2thehris commented:
Cleaned things up further and added tests to cover the missed lines.

Naming was a bit difficult, but I did my best. Happy to incorporate any suggestions.

friedemannsommer reviewed (approved):
looks good, thank you.

friedemannsommer commented:
I totally forgot to remind you to add a changeset.

Do you want to add it, or should I add it and give you credit?

C2thehris commented:
> I totally forgot to remind you to add a changeset.
> 
> Do you want to add it, or should I add it and give you credit?
> 
> If you decide to add it, please use "major" as the next version change, as new field names have been added and the `Variant` enum has been reordered.

Created a new PR with the changeset. This is my first time contributing to a project like this, so please do let me know if I missed anything!
